pkgs/_macro_host/lib/src/package_config.dart

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:dart_model/dart_model.dart';
+import 'package:package_config/package_config.dart' as package_config;
+
+/// Reads a package config to determine information about macros.
+class PackageConfig {
+  final Uri uri;
+
+  PackageConfig({required this.uri});
+
+  /// Checks whether [name] is a macro annotation.
+  ///
+  /// If so, returns the qualified name of the macro implementation.
+  ///
+  /// If not, returns `null`.
+  ///
+  /// This is a placeholder implementation until `language/3728` is
+  /// implemented. It expects macros to be marked by a comment in the
+  /// annotation package `pubspec.yaml` that looks like this:
+  ///
+  /// ```
+  /// # macro <annotation name> <implementation name>
+  /// ```
+  ///
+  /// For example:
+  ///
+  /// ```
+  /// # macro declare_x_macro.dart#DeclareX package:_test_macros/declare_x_macro.dart#DeclareXImplementation
+  /// ```
+  QualifiedName? lookupMacroImplementation(QualifiedName name) {
+    var packageName = name.uri;
+    if (packageName.startsWith('package:') && packageName.contains('/')) {
+      packageName = packageName.substring('package:'.length);
+      packageName = packageName.substring(0, packageName.indexOf('/'));
+    } else {
+      throw ArgumentError('Name must start "package:" and have a path: $name');
+    }
+    final packageConfig = package_config.PackageConfig.parseBytes(
+        File.fromUri(uri).readAsBytesSync(), uri);
+    final libraryQualifiedName =
+        name.string.substring(name.uri.indexOf('/') + 1);
+
+    final matchingPackage = packageConfig.packages
+        .where((p) => p.name == packageName)
+        .toList()
+        .singleOrNull;
+
+    if (matchingPackage == null) {
+      throw StateError(
+          'Package "$packageName" not found in package config at "$uri".');
+    }
+
+    // TODO(language/3728): read macro annotation identifiers from package
+    // config. Until then, check the pubsec, to simulate what that feature will
+    // do.
+    final packageUri = matchingPackage.root;
+    final pubspecUri = packageUri.replace(
+        pathSegments:
+            packageUri.pathSegments.followedBy(['pubspec.yaml']).toList());
```

Comment:
You should be able to just do:

```suggestion
    final pubspecUri = packageUri.resolve('pubspec.yaml');
```

Reply:
I definitely tried that! But now I try it again it works. Weird. I wonder how I screwed up :)

Thanks.

Reply:
Possibly you didn't have a trailing `/` originally for the root URI. Afaik though the package config spec requires this trailing slash.